### PR TITLE
Minimal uprops provider

### LINKED ISSUE #882: Import ICU4C binary properties data into ICU4X
Using the upropdump tool (unicode-org/icu#1741), ICU4C can emit data for Unicode properties as a .toml file. The next step is to consume that data. Binary properties are a good place to start.

### PR DESCRIPTION
**Pull request by iainireland**
Fixes #882

This is enough to let me write a [separate crate](https://github.com/iainireland/icu4x-js-regexp) that implements the binary properties needed for JS regexps. When issue #883 is done, I can extend that crate to also handle enumerated properties.

The .toml files were created using upropdump (unicode-org/icu#1741). I've only added the set of binary properties that JS supports for now.

### CODE REVIEW

File: provider/testdata/data/uprops/Bidi_C.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#
```
Comment (sffc) [inline provider/testdata/data/uprops/Bidi_C.toml:1]: These need copyright headers.  Running `upropdump` with `-c` should add the ICU4C copyright header.  That might be sufficient, but the copyright tool will probably complain that you need to add the ICU4X copyright header as well.  I think we should probably skip the files in this directory in the tool, such that the ICU4C copyright header by itself is sufficient.

File: provider/uprops/src/binary.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+use crate::upropdump_serde;
+use icu_provider::prelude::*;
+use icu_provider::iter::IterableDataProviderCore;
+use icu_uniset::provider::*;
+use icu_uniset::UnicodeSetBuilder;
+use std::borrow::Cow;
+use std::fs;
+use std::path::PathBuf;
+
+pub struct BinaryPropertiesDataProvider {
```
Comment (sffc) [inline provider/uprops/src/binary.rs:14]: Nit: Exported types need docs.

File: provider/uprops/src/binary.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+impl BinaryPropertiesDataProvider {
+    pub fn new(root_dir: PathBuf) -> Self {
+        BinaryPropertiesDataProvider { root_dir }
+    }
+    fn get_toml_as_string(&self, name: &str) -> Result<String, DataError> {
+        let mut path: PathBuf = self.root_dir.clone().join(name);
+        path.set_extension("toml");
+        fs::read_to_string(&path).map_err(DataError::new_resc_error)
```
Comment (sffc) [inline provider/uprops/src/binary.rs:25]: Nit: It's preferable if you attach the filename to the error.  Most other ICU4X components do this by having an `Error` specific to their crate that has an `Io` variant.

File: provider/uprops/src/binary.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    let provider = BinaryPropertiesDataProvider::new(root_dir);
+
+    let whitespace: UnicodeSet = provider
+        .load_payload(&DataRequest {
+            resource_path: ResourcePath {
+                key: key::WHITE_SPACE_V1,
+                options: ResourceOptions::default(),
+            },
+        })
+        .unwrap()
```
Comment (sffc) [inline provider/uprops/src/binary.rs:88]: Nit: prefer `.expect()` to `.unwrap()`

File: provider/uprops/src/binary.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        .take_payload()
+        .unwrap()
+        .get()
+        .clone()
+        .try_into()
+        .unwrap();
+
+    assert!(whitespace.contains(' '));
```
Comment (sffc) [inline provider/uprops/src/binary.rs:96]: Nit: Add a negative test (set does NOT contain) as well (and perhaps a few more positive tests)

File: provider/uprops/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+mod binary;
+mod upropdump_serde;
+
```
Comment (sffc) [inline provider/uprops/src/lib.rs:7]: Nit: Module-level docs (which will also become the README)

File: provider/uprops/src/upropdump_serde.rs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    #[derive(serde::Deserialize)]
+    pub struct Main {
+        pub unicode_set: Level1,
+    }
+}
+
+pub mod enumerated {
+    // TODO
```
Comment (sffc) [inline provider/uprops/src/upropdump_serde.rs:26]: Nit: Rather than a bare `// TODO`, add an issue number, like `// TODO(#883)`

**Review by sffc — CHANGES_REQUESTED** [review]

**Comment by Manishearth** [comment]
kicking CI after a branch protection change

File: provider/uprops/src/binary.rs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        req: &DataRequest,
+    ) -> Result<DataResponse<'d, 's, UnicodePropertyV1Marker>, DataError> {
+        let toml_str: String = self.get_toml_as_string(&req.resource_path.key.sub_category)?;
+        let toml_data: upropdump_serde::binary::Main =
+            toml::from_str(&toml_str).map_err(DataError::new_resc_error)?;
+
+        let mut builder = UnicodeSetBuilder::new();
+        for (start, end) in toml_data.unicode_set.data.ranges {
+            let start = std::char::from_u32(start).ok_or(DataError::InvalidPayload)?;
+            let end = std::char::from_u32(end).ok_or(DataError::InvalidPayload)?;
```
Comment (sffc) [inline provider/uprops/src/binary.rs:41]: Nit: Please make a separate `Error` type for this crate with more detailed errors, and then use `new_resc_error` to convert it to a DataError.  `DataError::InvalidPayload` is too generic to be helpful.

**Review by sffc — CHANGES_REQUESTED** [review]

File: provider/testdata/data/uprops/Bidi_C.toml (hunk 1 above)
Comment (iainireland) [reply provider/testdata/data/uprops/Bidi_C.toml:3]: I've added ICU4C copyright headers.

File: provider/uprops/src/lib.rs (hunk 6 above)
Comment (iainireland) [reply provider/uprops/src/lib.rs:25]: Spent a while poking at this. Am I correct in concluding that the README is created / committed as part of a CI task, or do I need to take some concrete action beyond adding the module-level docs?

File: provider/core/src/error.rs [hunk 9]
```diff
@@ -38,6 +38,10 @@ pub enum Error {
     #[displaydoc("Payload is missing")]
     MissingPayload,
 
+    /// The payload is invalid. This error is usually unexpected.
+    #[displaydoc("Payload is invalid")]
+    InvalidPayload,
+
```
Comment (sffc) [inline provider/core/src/error.rs:44]: Nit: If you aren't using `InvalidPayload` anymore, please delete it.

File: provider/uprops/src/binary.rs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    fn get_toml_data(&self, name: &str) -> Result<upropdump_serde::binary::Main, Error> {
+        let mut path: PathBuf = self.root_dir.clone().join(name);
+        path.set_extension("toml");
+        let toml_str = fs::read_to_string(&path).map_err(|e| Error::Io(e, path.clone()))?;
+        toml::from_str(&toml_str).map_err(|e| Error::Toml(e, path))
+    }
+}
+
+impl<'d> DataProvider<'d, UnicodePropertyV1Marker> for BinaryPropertiesDataProvider {
```
Comment (sffc) [inline provider/uprops/src/binary.rs:32]: Nit: Please use `'data` now instead of `'d`.

File: provider/uprops/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+//! [`BinaryPropertiesDataProvider`].
+//!
+//! **Important:** This data provider implementation is not optimized
+//! for production use.  It is much more efficient if you use
+//! [`FsDataProvider`] or [`StaticDataProvider`] instead.
+//!
+//! [`ICU4X`]: ../icu/index.html
+//! [`DataProvider`]: icu_provider::prelude::DataProvider
+//! [`FsDataProvider`]: ../icu_provider_fs/struct.FsDataProvider.html
+//! [`FsDataProvider`]: ../icu_provider_blob/struct.StaticDataProvider.html
```
Comment (sffc) [inline provider/uprops/src/lib.rs:18]: Nit:
```suggestion
//! [`StaticDataProvider`]: ../icu_provider_blob/struct.StaticDataProvider.html
```

File: provider/uprops/src/lib.rs (hunk 6 above)
Comment (sffc) [reply provider/uprops/src/lib.rs:24]: After writing the docs in *lib.rs*, run `cargo make generate-readmes` to convert it to README.md.  The CI will check that you ran the command and checked in the artifact (so long as the README file is present).

**Review by sffc — APPROVED** [review]
Looks good!  Next steps, which can be in follow-up PRs, will be:

1. Tie this in with `icu4x-datagen`
2. Demonstrate how to use the new data in `icu_uniset::props`

**Review by sffc — DISMISSED** [review]
LGTM but I think you probably need to merge main into your branch?  There are CI failures.

**Comment by iainireland** [comment]
Merged in main and updated the version in Cargo.toml / Cargo.lock. I think this is finally ready to go.

**Review by sffc — APPROVED** [review]

**Pull request merged**